File: frontend/src/api/evaluationMetrics.ts
import { api, ApiError } from "./client";

/**
 * 260718 设计文档 §7 (验证方式) evaluation-metrics adapter
 * (harness/api/evaluation_metrics.py, harness/evaluation_metrics/*).
 * Every metric mirrors the backend's `{value, numerator, denominator,
 * applicable, note}` shape - `applicable=false` means "not enough data to
 * compute this yet", never a fabricated 0/1 (harness/golden_set/metrics.py's
 * `_metric()` convention, reused here).
 */

export interface MetricResult {
  value: number | null;
  numerator: number;
  denominator: number;
  applicable: boolean;
  note: string;
}

export interface CoverageByClassEntry {
  strategyClass: string;
  status: "covered" | "excluded" | "missing";
  reason: string;
}

export interface CoverageCompletenessMetric extends MetricResult {
  coverageByClass: CoverageByClassEntry[];
}

export interface ReasonedNoveltyMetric extends MetricResult {
  novelGroundedGenes: string[];
}

export interface EvaluationMetricsSummary {
  designProjectId: string;
  process: {
    groundingRate: MetricResult;
    coverageCompleteness: CoverageCompletenessMetric;
  };
  capability: {
    screeningAbility: MetricResult;
    reasonedNovelty: ReasonedNoveltyMetric;
  };
  sanityCheck: {
    reproductionRate: MetricResult;
  };
}

export interface DesignProjectSummary {
  designProjectId: string;
  status: string;
  referenceDdrIds: string[];
  createdAt: number;
}

export interface ConsistencySample {
  sampleIndex: number;
  fallbackUsed: boolean;
  strategies: Array<{ strategyClass: string; mechanismTarget: string }>;
}

export interface ConsistencyByClass {
  strategyClass: string;
  sampleCount: number;
  convergence: number;
}

export interface ConsistencyRun {
  runId: string;
  designProjectId: string;
  nSamples: number;
  samples: ConsistencySample[];
  convergenceReport: {
    samplesWithOutput: number;
    samplesWithOutputRate: number;
    byStrategyClass: ConsistencyByClass[];
    note: string;
  };
  createdBy: string;
  createdAt: number;
}

interface RawMetricResult {
  value: number | null;
  numerator: number;
  denominator: number;
  applicable: boolean;
  note: string;
}

function toMetric(r: RawMetricResult): MetricResult {
  return { value: r.value, numerator: r.numerator, denominator: r.denominator, applicable: r.applicable, note: r.note ?? "" };
}

interface RawCoverageByClassEntry {
  strategy_class: string;
  status: "covered" | "excluded" | "missing";
  reason: string;
}

interface RawCoverageCompletenessMetric extends RawMetricResult {
  coverage_by_class: RawCoverageByClassEntry[];
}

interface RawReasonedNoveltyMetric extends RawMetricResult {
  novel_grounded_genes: string[];
}

interface RawEvaluationMetricsSummary {
  design_project_id: string;
  process: { grounding_rate: RawMetricResult; coverage_completeness: RawCoverageCompletenessMetric };
  capability: { screening_ability: RawMetricResult; reasoned_novelty: RawReasonedNoveltyMetric };
  sanity_check: { reproduction_rate: RawMetricResult };
}

function toSummary(r: RawEvaluationMetricsSummary): EvaluationMetricsSummary {
  return {
    designProjectId: r.design_project_id,
    process: {
      groundingRate: toMetric(r.process.grounding_rate),
      coverageCompleteness: {
        ...toMetric(r.process.coverage_completeness),
        coverageByClass: r.process.coverage_completeness.coverage_by_class.map((c) => ({
          strategyClass: c.strategy_class,
          status: c.status,
          reason: c.reason,
        })),
      },
    },
    capability: {
      screeningAbility: toMetric(r.capability.screening_ability),
      reasonedNovelty: {
        ...toMetric(r.capability.reasoned_novelty),
        novelGroundedGenes: r.capability.reasoned_novelty.novel_grounded_genes ?? [],
      },
    },
    sanityCheck: {
      reproductionRate: toMetric(r.sanity_check.reproduction_rate),
    },
  };
}

interface RawDesignProjectSummary {
  design_project_id: string;
  status: string;
  reference_ddr_ids: string[];
  created_at: number;
}

/** `GET /api/evaluation-metrics/by-project/{project_id}` (harness/api/evaluation_metrics.py::list_design_projects_for_project) - resolves the outer project id (the frontend route param) to its engineering-design project row(s), newest first. */
export async function listDesignProjectsForProject(projectId: string): Promise<DesignProjectSummary[]> {
  const raw = await api.get<{ design_projects: RawDesignProjectSummary[] }>(`/api/evaluation-metrics/by-project/${projectId}`);
  return raw.design_projects.map((r) => ({
    designProjectId: r.design_project_id, status: r.status, referenceDdrIds: r.reference_ddr_ids, createdAt: r.created_at,
  }));
}

/** `GET /api/evaluation-metrics/projects/{design_project_id}/summary` - null if the design project does not exist (404). */
export async function getMetricsSummary(designProjectId: string): Promise<EvaluationMetricsSummary | null> {
  try {
    const raw = await api.get<RawEvaluationMetricsSummary>(`/api/evaluation-metrics/projects/${designProjectId}/summary`);
    return toSummary(raw);
  } catch (e) {
    if (e instanceof ApiError && e.status === 404) return null;
    throw e;
  }
}

/** `POST /api/evaluation-metrics/projects/{design_project_id}/reference-ddr` - links this design project to its source paper's DDR id(s), enabling 合理新颖/复现率. */
export async function setReferenceDdr(designProjectId: string, ddrIds: string[]): Promise<string[]> {
  const raw = await api.post<{ reference_ddr_ids: string[] }>(`/api/evaluation-metrics/projects/${designProjectId}/reference-ddr`, { ddr_ids: ddrIds });
  return raw.reference_ddr_ids;
}

function toConsistencyRun(r: {
  run_id: string; design_project_id: string; n_samples: number;
  samples: Array<{ sample_index: number; fallback_used: boolean; strategies: Array<{ strategy_class: string; mechanism_target: string }> }>;
  convergence_report: { samples_with_output: number; samples_with_output_rate: number; by_strategy_class: Array<{ strategy_class: string; sample_count: number; convergence: number }>; note: string };
  created_by: string; created_at: number;
}): ConsistencyRun {
  return {
    runId: r.run_id, designProjectId: r.design_project_id, nSamples: r.n_samples,
    samples: r.samples.map((s) => ({ sampleIndex: s.sample_index, fallbackUsed: s.fallback_used, strategies: s.strategies.map((st) => ({ strategyClass: st.strategy_class, mechanismTarget: st.mechanism_target })) })),
    convergenceReport: {
      samplesWithOutput: r.convergence_report.samples_with_output,
      samplesWithOutputRate: r.convergence_report.samples_with_output_rate,
      byStrategyClass: r.convergence_report.by_strategy_class.map((c) => ({ strategyClass: c.strategy_class, sampleCount: c.sample_count, convergence: c.convergence })),
      note: r.convergence_report.note,
    },
    createdBy: r.created_by, createdAt: r.created_at,
  };
}

/** `POST /api/evaluation-metrics/projects/{design_project_id}/consistency-runs` - draws n_samples independent LLM samples of the same design task; each call is a real model request (see harness/evaluation_metrics/consistency_sampler.py), so keep n_samples modest. */
export async function runConsistencySample(designProjectId: string, nSamples: number, actorId: string): Promise<ConsistencyRun> {
  const raw = await api.post<Parameters<typeof toConsistencyRun>[0]>(`/api/evaluation-metrics/projects/${designProjectId}/consistency-runs`, { n_samples: nSamples, actor_id: actorId });
  return toConsistencyRun(raw);
}

/** `GET /api/evaluation-metrics/projects/{design_project_id}/consistency-runs` - most recent first. */
export async function listConsistencyRuns(designProjectId: string): Promise<ConsistencyRun[]> {
  const raw = await api.get<{ runs: Parameters<typeof toConsistencyRun>[0][] }>(`/api/evaluation-metrics/projects/${designProjectId}/consistency-runs`);
  return raw.runs.map(toConsistencyRun);
}

File: harness/api/evaluation_metrics.py
"""260718 设计文档 §7 (验证方式) evaluation-metrics API routes. Every route
calls the same `harness.evaluation_metrics` service functions the test suite
exercises; no business logic lives here (same convention as
`harness/api/engineering_design.py`).
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.orm import Session

from harness.api.deps import get_db_session
from harness.engineering_design.models import EngineeringDesignProject
from harness.evaluation_metrics import aggregator, consistency_sampler
from harness.evaluation_metrics.consistency_sampler import NoHandoffForProjectError

router = APIRouter(prefix="/api/evaluation-metrics", tags=["evaluation-metrics"])


def _consistency_run_dict(run) -> dict:
    return {
        "run_id": run.run_id, "design_project_id": run.design_project_id, "n_samples": run.n_samples,
        "samples": run.samples, "convergence_report": run.convergence_report, "created_by": run.created_by,
        "created_at": run.created_at,
    }


@router.get("/by-project/{project_id}")
def list_design_projects_for_project(project_id: str, session: Session = Depends(get_db_session)) -> dict:
    """Resolves the outer `projects.project_id` (what the frontend route
    carries) to its `EngineeringDesignProject` row(s) - every other
    engineering-design route is keyed by `design_project_id` directly, so the
    metrics dashboard needs this lookup to get started."""
    rows = session.execute(
        select(EngineeringDesignProject)
        .where(EngineeringDesignProject.project_id == project_id)
        .order_by(EngineeringDesignProject.created_at.desc())
    ).scalars().all()
    return {
        "design_projects": [
            {
                "design_project_id": r.design_project_id, "status": r.status,
                "reference_ddr_ids": r.reference_ddr_ids or [], "created_at": r.created_at,
            }
            for r in rows
        ]
    }


@router.get("/projects/{design_project_id}/summary")
def get_metrics_summary(design_project_id: str, session: Session = Depends(get_db_session)) -> dict:
    try:
        return aggregator.compute_all_metrics(session, design_project_id)
    except ValueError as e:
        raise HTTPException(404, str(e))


class ReferenceDdrBody(BaseModel):
    ddr_ids: list[str]


@router.post("/projects/{design_project_id}/reference-ddr")
def set_reference_ddr(design_project_id: str, body: ReferenceDdrBody, session: Session = Depends(get_db_session)) -> dict:
    proj = session.get(EngineeringDesignProject, design_project_id)
    if proj is None:
        raise HTTPException(404, f"no such design project: {design_project_id}")
    proj.reference_ddr_ids = body.ddr_ids
    session.flush()
    return {"design_project_id": design_project_id, "reference_ddr_ids": proj.reference_ddr_ids}


class ConsistencyRunBody(BaseModel):
    n_samples: int = consistency_sampler.DEFAULT_N_SAMPLES
    actor_id: str


@router.post("/projects/{design_project_id}/consistency-runs")
def start_consistency_run(design_project_id: str, body: ConsistencyRunBody, session: Session = Depends(get_db_session)) -> dict:
    try:
        run = consistency_sampler.run_consistency_sample(
            session, design_project_id=design_project_id, n_samples=body.n_samples, actor_id=body.actor_id,
        )
    except NoHandoffForProjectError as e:
        raise HTTPException(422, str(e))
    except ValueError as e:
        raise HTTPException(404, str(e))
    return _consistency_run_dict(run)


@router.get("/projects/{design_project_id}/consistency-runs")
def list_consistency_runs_route(design_project_id: str, session: Session = Depends(get_db_session)) -> dict:
    runs = consistency_sampler.list_consistency_runs(session, design_project_id)
    return {"runs": [_consistency_run_dict(r) for r in runs]}


@router.get("/projects/{design_project_id}/consistency-runs/{run_id}")
def get_consistency_run(design_project_id: str, run_id: str, session: Session = Depends(get_db_session)) -> dict:
    from harness.evaluation_metrics.models import ConsistencySamplingRun

    run = session.get(ConsistencySamplingRun, run_id)
    if run is None or run.design_project_id != design_project_id:
        raise HTTPException(404, f"no such consistency run: {run_id}")
    return _consistency_run_dict(run)
